dragonfly/implementations/psyllid_provider.py

```diff
@@ -0,0 +1,279 @@
+class MultiPsyllidProvider(core.Provider):
+    '''
+    Provider for direct communication with up to 3 Psyllid instances with a single stream each 
+    '''
+    def __init__(self,
+                 queue_a='channel_a_psyllid',
+                 queue_b = 'channel_b_psyllid',
+                 queue_c = 'channel_c_psyllid',
+                 set_condition_list = [],
+                 **kwargs):
+
+        core.Provider.__init__(self, **kwargs)        
+        self.queue_dict = {'a':queue_a, 'b':queue_b, 'c':queue_c}
+        self._set_condition_list = set_condition_list
+        self.status_dict = {'a':None, 'b':None, 'c':None}
+        self.status_value_dict = {'a':None, 'b':None, 'c':None}
+        self.channel_dict = {'a': 'ch0', 'b': 'ch1', 'c': 'ch2'}
+        self.freq_dict = {'a': None, 'b': None, 'c': None}
+        self.mode_dict = {'a':None, 'b':None, 'c':None}
+
+
+    def _finish_configure(self):
+        for channel in self.channel_dict.keys():
+            if self.request_status(channel)!=None:
+                self.mode_dict[channel]='streaming'
+
+                try: 
+                    if self.set_central_frequency(channel=channel, cf=800e6)==False:
+                       self.mode_dict[channel]='triggered'
+                       if self.set_central_frequency(channel=channel, cf=800e6)==False:
+                            self.mode_dict[channel]=None
+                            raise core.exceptions.DriplineGenericDAQError("Stream writer of Psyllid instance for channel {} is not configured correctly".format(channel))
+
+                    if self.get_number_of_streams(channel)!=1:
+                        self.mode_dict[channel]=None
+                        raise core.exceptions.DriplineGenericDAQError("Streams of Psyllid instance for channel {} are not configured correctly".format(channel))
+
+                except core.exceptions.DriplineGenericDAQError:
+                    logger.warning('No matching Psyllid instance for channel {} present'.format(channel))
+                except Exception as e:
+                    logger.error('Unexpected error during startup')
+                    raise 
+
+        # Summary after startup
+        logger.info('Status of channels: {}'.format(self.status_value_dict))
+        logger.info('Set central frequencies: {}'.format(self.freq_dict))
+        logger.info('Streaming or triggered mode: {}'.format(self.mode_dict))
+
+
+    def request_status(self, channel):
+        logger.info('Checking Psyllid status of channel {}'.format(channel))
+        try:
+            result = self.provider.get(self.queue_dict[channel]+'.daq-status', timeout=10)
+        except core.exceptions.DriplineError:
+            logger.warning('Psyllid instance for channel {} is not running or returned error'.format(channel))
+            self.status_dict[channel]=None
+            self.status_value_dict[channel]=None
+            return self.status_value_dict[channel]
```

Comment:
2.1) 
i) The reason for the else block is just being explicit.
ii) whenever the status of psyllid needs to be checked it is re-requested. 
The reason why the status can be none instead of raising an error in psyllid provider is that the psyllid provider class works with up to 3 psyllid instances. If you only start one psyllid instance for example the status for the other 2 is none because they are not there. This should not prevent psyllid_provider working normally with the one psyllid instance which is there.
The roach_daq_run_interface for a single channel requests the psyllid status for this channel via the psyllid_provider. if the status is none an error will be raised in the roach_daq_run_interface.